rditer: simplify rditer.Select

tbg opened the pull request:
Prompted by a recent PR discussion[^1].

`rditer.Select` and in particular the `ReplicatedSpansFilter` (some of the
options for which are additive while others are subtractive) makes it difficult
to see which key ranges are selected.

We simplify this by instead having a boolean for each span, and, over a few
commits, migrate every caller over into this new API.

----

I took the liberty to introduce the naming of a "system" span, which holds the
range descriptor, txn record, probes, and other less important keys. We
previously referred to this as "range local replicated key span" but there is of
other "range local replicated" data around (the rangeID-based replicated
keyspace) and I feel like this naming must be causing not only me recurring
periods of confusion.

The term "System" aligns with existing MVCC stats terminology, which already
categorizes these keys as "system" bytes.

[^1]: #147268 (comment)

Epic: CRDB-49111

tbg committed:
rditer: simplify rditer.Select

Prompted by a recent PR discussion[^1].

`rditer.Select` and in particular the `ReplicatedSpansFilter` (some of the
options for which are additive while others are subtractive) makes it difficult
to see which key ranges are selected.

We simplify this by instead having a boolean for each span, and, over a few
commits, migrate every caller over into this new API.

----

I took the liberty to introduce the naming of a "system" span, which holds the
range descriptor, txn record, probes, and other less important keys. We
previously referred to this as "range local replicated key span" but there is of
other "range local replicated" data around (the rangeID-based replicated
keyspace) and I feel like this naming must be causing not only me recurring
periods of confusion.

The term "System" aligns with existing MVCC stats terminology, which already
categorizes these keys as "system" bytes.

[^1]: https://github.com/cockroachdb/cockroach/pull/147268#discussion_r2107080717

tbg committed:
engineccl: refactor

tbg committed:
kvserver: refactor clearSubsumedReplicaDiskData

tbg committed:
gc: refactor

tbg committed:
kvstorage: refactor

tbg committed:
kvserver: refactor

tbg committed:
rditer: adopt new API

tbg committed:
rditer: prepare IterateReplicaKeySpans for refactor

It now accepts the old filter or an explicit selection of what to iterate over.
Follow-up commits will phase out the filter.

tbg committed:
rditer: shift ReplicatedSpansFilter enum by one

tbg committed:
rditer: use updated IterateReplicaKeySpans for tests

Note that this exercises all combination of replicated spans now, and thus
created new files. It did not mutate any of the existing files, i.e. the output
for those cases did not change.

tbg committed:
rditer: remove testdata legacy names

Lots of file changes, but it's clear that this is just a mechanical change
since no business logic was changed in this commit other than the test naming.

tbg committed:
rditer: push SelectOpts into IterateReplicaKeySpans

This allows us to phase out the unreplicatedOnly bool with each caller
migration.  Note that some datadriven tests changed because the test was
updated and no longer disables the rangeID-replicated key space in the case in
which only the user replicated span is requested. The callers that want this
behavior still get it since they are not migrated, and we will set it via
SelectOpts when they are migrated.

tbg committed:
kvserver: migrate kvSnapshotStrategy.Send

This is the only significant caller of ReplicatedSpansUserOnly.
A test usage remains, to be removed separately.

tbg committed:
rditer: migrate TestSelect

tbg committed:
rditer: remove ReplicatedSpansUserOnly

tbg committed:
rditer: phase out ReplicatedSpansAll

tbg committed:
rditer: remove ReplicatedSpansExcludeUser

tbg committed:
rditer: remove ReplicatedSpansExcludeLocks

tbg committed:
rditer: remove ReplicatedSpansLocksOnly

tbg committed:
rditer: remove ReplicatedSpansFilter

tbg committed:
rditer: remove ReplicatedBySpan field

tbg committed:
rditer: remove replicatedOnly arg

tbg committed:
rditer: remove another replicatedOnly arg

tbg commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -19,9 +19,12 @@ import (
 const (
+	// Throw away the zero value for easier migration away from this deprecated
+	// field.
+	_ ReplicatedSpansFilter = iota
```
I just moved this out of the way so that the zero value is the "we're not using the legacy option" value.
This type is gone by the end of this PR.

I don't want to speculate what the old code would have done with an undefined value. I could look at it, but I just spent a few hours getting rid of it and don't remember it at this point 😆

pav-kv reviewed (commented):
Dumping review progress so far. Couldn't make it to the end today, will continue tomorrow.

pav-kv commented on pkg/kv/kvserver/rditer/replica_data_iter_test.go:
```diff
@@ -283,27 +276,46 @@ func TestReplicaDataIterator(t *testing.T) {
-					}
-					t.Run(fmt.Sprintf("replicatedSpans=%v", replicatedKeysName), func(t *testing.T) {
-						name := "all"
+				// Test all combinations of boolean fields.
```
Why do we need the combinatorics in a datadriven test? Reviewing all these files is infeasible and nobody will do.

I think what's useful is to have one "select all" test, and/or one test which does mini-select for each of the possible sub-spans. This serves as a documentation + golden test.

The combinatorics of things can then be tested with a regular unit-test. We want 2 properties: `select(A|B) == select(A)|select(B)`, and `select(anything)` is sorted by key. These kinds of properties/invariants are better to check in code than by eyeballing.

pav-kv replied:
Demoed this on `TestSelect` in the `rditer: improve TestSelect` commit.

pav-kv committed:
kvserver,cli: idiomatically inline Select params

Epic: none
Release note: none

pav-kv committed:
rditer: rename Span to RSpan

Epic: none
Release note: none

pav-kv committed:
rditer: index keys covered by SelectRangedOptions

Epic: none
Release note: none

pav-kv commented on pkg/kv/kvserver/rditer/select.go:
```diff
@@ -39,6 +40,43 @@ const (
+type SelectRangedOptions struct {
+	// Span describes the range bounds. This must be set if any of the other
+	// fields in this struct are.
+	Span roachpb.RSpan
```
Done.

pav-kv commented on pkg/kv/kvserver/rditer/replica_data_iter.go:
```diff
@@ -425,42 +425,29 @@ func IterateReplicaKeySpans(
+	// NB: ignored if SelectOpts passed as filterOrOptions.
 	replicatedOnly bool, // TODO(tbg): remove when SelectRangedOptions is used
-	filterOrOptions interface{}, // ReplicatedSpansFilter or SelectRangedOptions
+	filterOrOptions interface{}, // ReplicatedSpansFilter or SelectOpts
```
Powered through by just reviewing the end state.

pav-kv committed:
rditer: improve TestSelect

Minimize the datadriven output for human review, and improve the
assertion power of the exhaustive combinatorics in the test code.

Epic: none
Release note: none

arulajmani reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I mostly just reviewed for the before vs. after as the code evolved.

> Reviewed 13 of 13 files at r1, 11 of 11 files at r3, 12 of 12 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, 1 of 1 files at r11, 31 of 31 files at r12, 19 of 19 files at r13, 9 of 9 files at r14, 1 of 1 files at r15, 8 of 8 files at r16, 1 of 1 files at r17, 5 of 5 files at r18, 1 of 1 files at r19, 1 of 1 files at r20, 1 of 1 files at r21, 10 of 10 files at r22, 10 of 10 files at r23, 7 of 7 files at r24, 1 of 1 files at r25, 3 of 3 files at r26, 17 of 17 files at r27, 1 of 1 files at r28.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147295#-:-ORTgY717H8pEP-zY8UO:b-f6bwcy)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens, @pav-kv, and @tbg)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147295#-:-ORThvrz7UxRzQVN-1pd:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens, @pav-kv, and @tbg)



<!-- Sent from Reviewable.io -->

pav-kv commented:
bors r+
